cli/args/mod.rs

```diff
@@ -895,50 +870,78 @@ impl CliOptions {
       include_task_comments: matches!(
         flags.subcommand,
         DenoSubcommand::Task(..)
       ),
     };
-    let maybe_config_file = ConfigFile::discover(
-      &flags.config_flag,
-      flags.config_path_args(&initial_cwd),
-      &initial_cwd,
-      additional_config_file_names,
-      &parse_options,
-    )?;
+    let discover_pkg_json = flags.config_flag
+      != deno_config::ConfigFlag::Disabled
+      && !flags.no_npm
+      && !has_flag_env_var("DENO_NO_PACKAGE_JSON");
+    if !discover_pkg_json {
+      log::debug!("package.json auto-discovery is disabled");
+    }
 
-    let mut maybe_package_json = None;
-    if flags.config_flag == deno_config::ConfigFlag::Disabled
-      || flags.no_npm
-      || has_flag_env_var("DENO_NO_PACKAGE_JSON")
-    {
-      log::debug!("package.json auto-discovery is disabled")
-    } else if let Some(config_file) = &maybe_config_file {
-      let specifier = config_file.specifier.clone();
-      if specifier.scheme() == "file" {
-        let maybe_stop_at = specifier
-          .to_file_path()
-          .unwrap()
-          .parent()
-          .map(|p| p.to_path_buf());
-
-        maybe_package_json =
-          discover_package_json(&flags, maybe_stop_at, &initial_cwd)?;
+    let workspace = match &flags.config_flag {
+      deno_config::ConfigFlag::Discover => {
+        if let Some(start_dirs) = flags.config_path_args(&initial_cwd) {
+          Workspace::discover(
+            WorkspaceDiscoverStart::Dirs(&start_dirs),
+            &WorkspaceDiscoverOptions {
+              fs: &DenoConfigFsAdapter::new(&RealFs),
+              pkg_json_cache: Some(
+                &deno_runtime::deno_node::PackageJsonThreadLocalCache,
+              ),
+              config_parse_options,
+              additional_config_file_names,
+              discover_pkg_json,
+            },
+          )?
+        } else {
+          Workspace::empty(Arc::new(
+            ModuleSpecifier::from_directory_path(&initial_cwd).unwrap(),
+          ))
+        }
       }
-    } else {
-      maybe_package_json = discover_package_json(&flags, None, &initial_cwd)?;
+      deno_config::ConfigFlag::Path(path) => {
+        let config_path = normalize_path(initial_cwd.join(path));
+        Workspace::discover(
+          WorkspaceDiscoverStart::ConfigFile(&config_path),
+          &WorkspaceDiscoverOptions {
+            fs: &DenoConfigFsAdapter::new(&RealFs),
+            pkg_json_cache: Some(
+              &deno_runtime::deno_node::PackageJsonThreadLocalCache,
+            ),
+            config_parse_options,
+            additional_config_file_names,
+            discover_pkg_json,
+          },
+        )?
+      }
+      deno_config::ConfigFlag::Disabled => Workspace::empty(Arc::new(
+        ModuleSpecifier::from_directory_path(&initial_cwd).unwrap(),
+      )),
+    };
+
+    for diagnostic in workspace.diagnostics() {
+      if diagnostic.kind.is_fatal() {
+        return Err(diagnostic.into());
+      }
```

Comment:
Maybe collect all diagnostics and somehow concat them? Otherwise it might require a big back and forth if they are printed one by one. Also is the order deterministic?

Reply:
I think it's maybe not worth it. These diagnostics will be rare and currently the only fatal one is when a workspace member has the same name as another one. Actually, I should probably just get rid of the fatal daignostics and surface an error instead.

Reply:
https://github.com/denoland/deno_config/pull/68